compiler/src/Language/Bond/Codegen/Java/Enum_java.hs

```diff
@@ -30,37 +30,73 @@ package #{javaPackage};
 
     typeDefinition Enum {..} = [lt|
 #{Java.generatedClassAnnotations}
-public class #{declName} {
-    #{newlineSep 1 constant enumConstantsWithInt}
+public final class #{declName} {
+
+    public static final class Values {
+        private Values() {}
+
+        #{newlineSep 2 constantIntValueDecl enumConstantsWithInt}
+    }
+
+    #{newlineSep 1 constantObjectDecl enumConstantsWithInt}
 
     public final int value;
 
     public #{declName}(int value) { this.value = value; }
 
     @Override
-    public boolean equals(Object other) {
-        if (!(other instanceof #{declName})) { return false; }
-        return this.value == ((#{declName}) other).value;
-    }
+    public boolean equals(Object other) { return (other instanceof #{declName}) && (this.value == ((#{declName}) other).value); }
+ 
+    @Override
+    public int hashCode() { return this.value; }
 
     @Override
-    public int hashCode() {
-        return this.value;
+    public String toString() { return String.valueOf(this.value); }
+
+    public static #{declName} get(int value) {
+        switch (value) {
+            #{newlineSep 3 switchCaseConstantMapping enumConstantsWithIntDistinct}
+            default: return new #{declName}(value);
+        }
     }
 }|]
       where
-        -- constant
-        constant Constant {..} = let value x = [lt|#{x}|] in
-            [lt|public static final #{declName} #{constantName} = new #{declName}(#{optional value constantValue});|]
+        -- constant object
+        constantObjectDecl c@Constant {..} =
+            [lt|public static final #{declName} #{constantName} = #{enumObjectAssigmentValue c};|]
 
-        -- Process constants to make sure all values can be converted
-        -- to integer.
-        enumConstantsWithInt = fixEnumWithInt 0 enumConstants []
+        -- constant int
+        constantIntValueDecl Constant {..} = let value x = [lt|#{x}|] in
+            [lt|public static final int #{constantName} = #{optional value constantValue};|]
```

Comment:
I see - we rewrite them down there so they all have a value. If my understanding is correct, this is fine, though I'd be more jazzed if we errored out right here if a `Nothing` ever made it through.